### PR TITLE
Improve frontend preview

### PR DESCRIPTION
**Pull request by aschempp**
This adjusts the following:

 1. the JwtToken is parsed in all entry points
 2. To enable the debug mode, the `preview.php` is not necessary (therefore we no longer require the back end to switch the entry point)
 3. If the JwtManager is not set in the container (manager-bundle is not installed), do not show the "debug" button in the back end
 4. If the `APP_DEBUG` env variable is set, do not load the JwtManager (meaning debug cannot be enabled or disabled from the back end)
 5. Make the preview entry point script configurable in the container. For a regular Sf App, the preview entry point is empty, so preview will be on the regular entry point. I don't see this as a problem, as there might not be a reverse proxy at all.

**TODO:**
 - [x] Force back end login on preview entry point
 - [x] Update unit tests

### CODE REVIEW

File: core-bundle/src/Resources/contao/controllers/BackendMain.php [hunk 1]
```diff
@@ -95,21 +94,12 @@ public function __construct()
...
-
-				$this->redirect($script . $objRequest->getPathInfo() . $qs);
-			}
+			// This will throw an exception if the JwtManager does not exist (not a Managed Edition).
+			// As we do not show the debug button in that case, this should never be reached.
+			$objJwtManager = System::getContainer()->get('contao_manager.jwt_manager');
 
 			$strReferer = Input::get('referer') ? '?' . base64_decode(Input::get('referer', true)) : '';
-			$objResponse = new RedirectResponse($script . $objRequest->getPathInfo() . $strReferer);
+			$objResponse = new RedirectResponse($objRequest->getPathInfo() . $strReferer);
```
Comment (Toflar) [inline core-bundle/src/Resources/contao/controllers/BackendMain.php]: So this has two implications:

1. If you do not have a ME, you cannot have the debug mode enabled without implementing the `contao_manager.jwt_manager` service yourself? Maybe we could improve that?

2. If you're not logged in, have a cache entry, and enable the debug mode, you'll never see it...

Comment (aschempp) [reply core-bundle/src/Resources/contao/controllers/BackendMain.php]: 1. If you do not have a managed edition, I don‘t think a back end user should decide on debug mode. You would run a manual AppKernel and use environment variables. 

2. that could be correct, for the front end. But if you use the preview button, everything works. And you also have a JWT cookie, so your cache will be bypassed in most cases.

Comment (Toflar) [reply core-bundle/src/Resources/contao/controllers/BackendMain.php]: 1. Fair point.
2. You have a cookie, so it's bypassed. True.

**Comment by Toflar** [comment]
Concept LGTM 👍

**Comment by leofeyer** [comment]
> Make the preview entry point script configurable in the container. For a regular Sf App, the preview entry point is empty, so preview will be on the regular entry point. I don't see this as a problem, as there might not be a reverse proxy at all.

I would prefer a unified solution for both Varnish and the Symfony HTTP cache.

Also, we will get rid of the `contao/preview` route as soon as we have an inline preview bar and then its nice if we can prepend the `/preview.php` fragment to any URL to enable it, isn't it?

* `https://domain.tld/news.html`
* `https://domain.tld/preview.php/news.html`

**Review by leofeyer — CHANGES_REQUESTED** [review]
I like the concept and I think it is a great improvement. 👍 

However, I would prefer a unified solution for both Varnish and the Symfony HTTP cache, which means we must always use the `preview.php` entry point for the FE preview.

* `https://domain.tld/news.html`
* `https://domain.tld/preview.php/news.html`

**Comment by aschempp** [comment]
what makes you think we are _not_ always using the `preview.php`?

**Comment by leofeyer** [comment]
The fact that `contao.preview_script` is an empty string by default?

**Comment by Toflar** [comment]
That's why in the ME it's configured to `preview.php`? 😄 https://github.com/contao/contao/pull/592/files#diff-45904db91fdd5ee191427694f05cbd17R10

**Comment by leofeyer** [comment]
Yes, but I don't see why it has to be configurable at all. Contao uses the `preview.php` entry point to generate the front end preview, so if you want to use Contao, you have to install the `preview.php` file. Why does it have to be renamable?

**Comment by aschempp** [comment]
Because if you use Contao as a Symfony Bundle, we must not force anything, because we can‘t make any assumption. 

If you do not have a cache at all, you don‘t need a separate entry point. If your entry point bypasses the cache for your business IP range, you don‘t need a separate entry point.

I agree there‘s a slim chance. But we also make the web folder configurable, and the image cache folder etc.
I don‘t see why we wouldn‘t do it here

**Comment by leofeyer** [comment]
Valid point. 👍

Can I do anything to help you get this PR from "Draft" to "RTM"?

**Comment by aschempp** [comment]
Unfortunately, it is currently not possible to redirect to the requested front end page, as the back end redirect only supports query parameters (and no other routes). Maybe we can improve on that (in a separate PR imho)

File: manager-bundle/bin/contao-console [hunk 2]
```diff
@@ -31,12 +31,8 @@ if (file_exists(__DIR__.'/../autoload.php')) {
 $loader = require $projectDir.'/vendor/autoload.php';
 
 $input = new ArgvInput();
-$env = $input->getParameterOption(['--env', '-e'], getenv('SYMFONY_ENV') ?: 'prod');
+$env = $input->getParameterOption(['--env', '-e'], null);
```
Comment (leofeyer) [inline manager-bundle/bin/contao-console]: This seems strange to me, as the interface expects a boolean value as second argument.

File: manager-bundle/src/HttpKernel/ContaoKernel.php [hunk 3]
```diff
@@ -216,16 +235,20 @@ public static function setProjectDir(string $projectDir): void
...
 
-    public static function create(string $projectDir, bool $debug = false): self
+    public static function create(string $projectDir, string $env = null): self
     {
-        if (file_exists($projectDir.'/.env')) {
-            (new Dotenv(false))->load($projectDir.'/.env');
+        self::loadEnv($projectDir);
+
+        if (null === $env) {
+            $env = (string) ($_SERVER['APP_ENV'] ?? $_SERVER['SYMFONY_ENV'] ?? 'prod');
```
Comment (leofeyer) [inline manager-bundle/src/HttpKernel/ContaoKernel.php:243]: `SYMFONY_ENV` is the legacy key from the pre-flex era. Do we really want to support it?

File: manager-bundle/src/HttpKernel/ContaoKernel.php [hunk 4]
```diff
@@ -216,16 +235,20 @@ public static function setProjectDir(string $projectDir): void
...
 
-        // TODO: use manager config to load settings
+        if ('dev' !== $env && 'prod' !== $env) {
+            die('The Contao Managed Edition only supports "dev" and "prod" environment.');
+        }
 
-        // See https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/3.3/public/index.php#L27
-        if ($trustedProxies = $_SERVER['TRUSTED_PROXIES'] ?? null) {
+        // See https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/4.2/public/index.php
+        if ($trustedProxies = $_SERVER['TRUSTED_PROXIES'] ?? false) {
```
Comment (leofeyer) [inline manager-bundle/src/HttpKernel/ContaoKernel.php]: `bool` is not a valid type for `$_SERVER['TRUSTED_PROXIES']`.

File: manager-bundle/src/HttpKernel/ContaoKernel.php [hunk 5]
```diff
@@ -311,4 +364,31 @@ private function addBundlesFromPlugins(array &$bundles): void
...
+
+        $dotEnv = new Dotenv(false);
+
+        if (method_exists($dotEnv, 'loadEnv')) {
+            $dotEnv->loadEnv($projectDir.'/.env', $varName, 'prod', []);
+
+            return;
+        }
+
+        // BC for Symfony <= 4.1 where Dotenv::loadEnv does not exist
```
Comment (leofeyer) [inline manager-bundle/src/HttpKernel/ContaoKernel.php]: Our minimum Symfony version is 4.2, so this code can probably be removed?

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/EventListener/PreviewAuthenticationListener.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+        if (
+            '' === $this->previewScript
+            || $request->getScriptName() !== $this->previewScript
+            || !$this->scopeMatcher->isFrontendRequest($request)
+        ) {
+            return;
+        }
+
+        if (!$this->tokenChecker->hasBackendUser()) {
```
Comment (leofeyer) [inline core-bundle/src/EventListener/PreviewAuthenticationListener.php]: Why did you not add the condition above?

```php
        if (
            '' === $this->previewScript
            || $request->getScriptName() !== $this->previewScript
            || !$this->scopeMatcher->isFrontendRequest($request)
            || $this->tokenChecker->hasBackendUser()
        ) {
            return;
        }

        $event->setResponse(new RedirectResponse($this->router->generate('contao_backend_login')));
    }
```

**Comment by leofeyer** [comment]
> as the back end redirect only supports query parameters (and no other routes)

What exactly do you mean?

**Review by Toflar — CHANGES_REQUESTED** [review]
- [x] Login and logout should not set/unset the JWT
- [x] Shouldn't the JWT be part of the core-bundle? Either that or the complete debug functionality has to be provided by the manager-bundle (so the button should be added dynamically, language files etc.). Right now, core-bundle and manager-bundle responsibilities are mixed.
- [x] The `PreviewAuthenticationListener` should be part of the manager-bundle.

**Comment by leofeyer** [comment]
> Shouldn't the JWT be part of the core-bundle?

I don't think so, as the whole debug logic is only part of the managed edition. A regular Symfony application uses environment variables.

However, moving the whole functionality requires to add the manager bundle on Transifex and to use the KNP menu bundle for the back end top navigation – both can only be done in Contao 4.9 at this point.

> The `PreviewAuthenticationListener` should be part of the manager-bundle

Why do you think so? The front end preview is not limited to the managed edition, is it?

**Comment by aschempp** [comment]
> > The `PreviewAuthenticationListener` should be part of the manager-bundle
> 
> Why do you think so? The front end preview is not limited to the managed edition, is it?

Because the Managed Edition decides whether that preview entry point should only be accessible to logged in users. Maybe a regular application has an IP protected preview endpoint that does not need Contao authentication. Implemented in 61ca51cbcbfd09f90553285b4e401e0fd9384341

**Comment by aschempp** [comment]
I have now moved everything possible to the manager-bundle. That includes the labels, but I'm not sure if adding them on Transifex will be an issue for 4.8?

**Comment by leofeyer** [comment]
I can add them and thanks to translation memory, Transifex should find the translations automatically.

**Review by Toflar — APPROVED** [review]

**Review by leofeyer — COMMENTED** [review]
Very good work, thank you @aschempp. There are only two things that bother me:

1. Do we really want to support the legacy `SYMFONY_ENV` variable?

2. Is it necessary to put the translations into the old `contao/languages` folder? I would prefer to have them in the `translations` folder like we do in the [installation-bundle](https://github.com/contao/contao/tree/master/installation-bundle/src/Resources/translations).

**Comment by aschempp** [comment]
> 1. Do we really want to support the legacy `SYMFONY_ENV` variable?

Is it really legacy? If I want to globally set this variable on my local computer, I would very much prefer `SYMFONY_ENV` over `APP_ENV`. Also, removing it would be kind of a BC break 🤷‍♂ 

> 2. Is it necessary to put the translations into the old `contao/languages` folder? I would prefer to have them in the `translations` folder like we do in the [installation-bundle](https://github.com/contao/contao/tree/master/installation-bundle/src/Resources/translations).

We could do that, it would probably actually be better. But I wonder why you're using the messages domain, shouldn't we use a domain for the bundle name? Shall I update the PR?

**Comment by leofeyer** [comment]
> If I want to globally set this variable on my local computer, I would very much prefer `SYMFONY_ENV` over `APP_ENV`.

It does not matter what we would prefer. Symfony 4 does not handle `SYMFONY_ENV` anymore, so we should not, either, now that Symfony 4.2 is our minimum version.

> But I wonder why you're using the messages domain, shouldn't we use a domain for the bundle name?

It is kind of the default domain, isn't it?

> Shall I update the PR?

Yes please.

**Comment by aschempp** [comment]
> > If I want to globally set this variable on my local computer, I would very much prefer `SYMFONY_ENV` over `APP_ENV`.
> 
> It does not matter what we would prefer. Symfony 4 does not handle `SYMFONY_ENV` anymore, so we should not, either, now that Symfony 4.2 is our minimum version.

Well in Symfony you can adjust it to your liking, you can't do that in Contao. And having a global `APP_ENV` variable is really not nice. Also, as I said, it would be kinda a BC break.

> > But I wonder why you're using the messages domain, shouldn't we use a domain for the bundle name?
> 
> It is kind of the default domain, isn't it?

It's the default domain for your app! Not for bundles, afaik. Maybe @Toflar / @sheeep / @bytehead can shed more light on this.

**Comment by Toflar** [comment]
Yes, you should use your own domain for message translations. That's the whole point of a domain. E.g. in Symfony if you want to extend or adjust the validation messages, you have to use the `validator` domain. The FOSUserBundle for example uses `FOSUserBundle` (https://github.com/FriendsOfSymfony/FOSUserBundle/tree/master/Resources/translations) but I'd say we should just use the same key as we would use to configure the bundle. So for Contao that'd be `contao`, if I were to have translations for my Bynder integration I'd name the domain `terminal42_conao_bynder` (https://github.com/terminal42/contao-bynder/blob/master/src/DependencyInjection/Configuration.php#L25) etc.

**Comment by leofeyer** [comment]
> Also, as I said, it would be kinda a BC break.

There certainly is no BC promise for renamed environment variables. 😄 

> It's the default domain for your app! Not for bundles, afaik.

I see. Would you prefer to use separate domains per bundle then (`contao_manager` and `contao_installation`) or one `contao` domain for all bundles in `contao/contao`?

**Comment by ausi** [comment]
All message domains beginning with `contao_` are reserved for our decorated translator...

**Comment by aschempp** [comment]
> All message domains beginning with contao_ are reserved for our decorated translator...

😮 you're absolutely right, I forgot about this! I can only see three options then:
 1. Use `contao` for all our bundles. Not really nice, probably also harder to sync with Transifex and harder for people to realize where the translations are from
 2. Use `ContaoCoreBundle` / `ContaoManagerBundle` like FOSUserBundle does. Not really nice names for me though, I agree with @Toflar they should be the bundle alias and they should be lowercase (dunno why, just a gut feeling).
 3. Add an explicit except for `contao_installation`, `contao_manager` (and possibly other bundles later) to our translator decorator.

Honestly, 3. is ugly but my preferred one …

File: core-bundle/tests/Translation/TranslatorTest.php [hunk 7]
```diff
@@ -56,14 +59,35 @@ public function testForwardsTheMethodCallsToTheDecoratedTranslator(): void
...
+
+        yield ['contao_installation'];
+
+        yield ['contao_listing'];
+
+        yield ['contao_manager'];
+
+        yield ['contao_news'];
+
+        yield ['contao_newsletter'];
```
Comment (leofeyer) [inline core-bundle/tests/Translation/TranslatorTest.php:80]: We should probably also test that `contao_default` and `contao_tl_page` are not forwarded.

Comment (aschempp) [reply core-bundle/tests/Translation/TranslatorTest.php:80]: We should add a lot of tests here, but that's not really related? But feel free to add 🙃

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thanks a lot @aschempp.

**Pull request merged**